### PR TITLE
gl_engine: implement advance blending

### PR DESCRIPTION
**Pull request by RuiwenTang**
Implement some advance blending equation in GLEngine by using two RenderPass.

* The first pass render the content into an off-screen framebuffer, also blit the screen content to an off-screen Texture.

* The second pass renders the final blended color to the screen. Also use a stencil test to prevent non-drawn areas from participating in the color blending calculation.

----------
Now I think the **Blending** Example is correct:
截屏2024-09-14 23 36 20

issue: https://github.com/thorvg/thorvg/issues/2435

### CODE REVIEW

File: src/renderer/gl_engine/tvgGlShaderSrc.cpp [hunk 1]
```diff
@@ -509,3 +509,90 @@ const char* BLIT_FRAG_SHADER = TVG_COMPOSE_SHADER(
         FragColor = texture(uSrcTexture, vUV);                      \n
     }
 );
+
+const char* COMPLEX_BLEND_FRAG_SHADER = TVG_COMPOSE_SHADER(
```
Comment (SergeyLebedkin) [inline src/renderer/gl_engine/tvgGlShaderSrc.cpp:513]: Is it possible too separate blend methods by the different fragment shaders insted of cheking for each pixel?

File: src/renderer/gl_engine/tvgGlRenderPass.h [hunk 2]
```diff
@@ -74,6 +74,10 @@ class GlRenderPass
     }
 
     int nextDrawDepth() { return ++mDrawDepth; }
+
+    void setDrawDepth(int32_t depth) { mDrawDepth = depth; }
+
```
Comment (SergeyLebedkin) [inline src/renderer/gl_engine/tvgGlRenderPass.h:79]: What is global purpose of the depth buffer in rendering pipelines in gl_engine?

Comment (RuiwenTang) [reply src/renderer/gl_engine/tvgGlRenderPass.h:79]: It is used for path clip.

**Comment by hermet** [comment]
Replaced due to a conflict and shader compilation errors: https://github.com/thorvg/thorvg/commit/d1af484bf9f85d4dc2556bc3742793b45b675879

@RuiwenTang I appreciate your hard work. It would be helpful if you use a branch directly from the ThorVG project (instead of a fork) so that I can update the PR directly when necessary. Thanks!

**Comment by RuiwenTang** [comment]
> Replaced due to a conflict and shader compilation errors: [d1af484](https://github.com/thorvg/thorvg/commit/d1af484bf9f85d4dc2556bc3742793b45b675879)
> 
> @RuiwenTang I appreciate your hard work. It would be helpful if you use a branch directly from the ThorVG project (instead of a fork) so that I can update the PR directly when necessary. Thanks!

I didn’t realize I had push permissions to the main repository, so I submitted PRs in the forked repository every time.
BTW: I noticed that this submission used my business email address. Can this be changed?

**Comment by hermet** [comment]
@RuiwenTang updated, please check it.

**Comment by hermet** [comment]
@RuiwenTang 

> I didn’t realize I had push permissions to the main repository,

Of course, you have it—you are the official maintainer(committer) of ThorVG.

**Comment by RuiwenTang** [comment]
I saw it. Thank you very much. @hermet

**Pull request closed without merging**